components/bin/build

```diff
@@ -0,0 +1,318 @@
+    const name = path.parse(file).name;
+    const lines = [
+        '"use strict";',
+        `Object.defineProperty(exports, '__esModule', {value: true});`
+    ];
+    let source = (dir.replace(/\//g, '.') + '.' + name).replace(/^\./, '')
+        .replace(/\.[^.]*/g, (x) => (x.substr(1).match(/[^a-zA-Z_]/) ? '[\'' + x.substr(1) + '\']' : x));
+    if (exists(path.resolve(MATHJAX3, file.replace(/\.ts$/, '')))) source += '_ts';
+    for (const id of objects) {
+        lines.push(`exports.${id} = MathJax._.${source}.${id};`);
+    }
+    return lines;
+}
+
+/**
+ * @param {string} file     Path to a file
+ * @return {boolean}        True if the file exists, false if not
+ */
+function exists(file) {
+    if (!fs.existsSync(file)) return false;
+    //
+    // For case-insensitive filing systems (like Mac OS),
+    //   check that the file names match exactly
+    //
+    const [dir, name] = [path.dirname(file), path.basename(file)];
+    return fs.readdirSync(dir).indexOf(name) >= 0;
+}
+
+/**
+ * Recursively make any needed directories
+ *
+ * @param {string} dir   The directory relative to the library directory
+ */
+function makeDir(dir) {
+    const fulldir = path.resolve(LIB, dir);
+    if (fs.existsSync(fulldir)) return;
+    makeDir(path.dirname(fulldir));
+    fs.mkdirSync(fulldir);
+}
+
+/**
+ * Make a file in the library directory from the given lines of javascript
+ *
+ * @param {string} file     The name of the file relative to the root .ts directory
+ * @param {string[]} lines  The contents of the file to create
+ */
+function makeFile(file, lines) {
+    if (!lines.length) return;
+    const [dir, name] = [path.dirname(file), path.basename(file)];
+    makeDir(dir);
+    fs.writeFileSync(path.resolve(LIB, dir, name.replace(/\.ts$/, '.js')), lines.join('\n') + '\n');
+}
+
+/**
+ * Make the library file that adds all the exported objects into the MathJax global object
+ */
+function processGlobal() {
+    console.log(COMPONENT + '.ts');
+    const lines = [
+        '"use strict";',
+        `Object.defineProperty(exports, '__esModule', {value: true});`,
+        `require('${GLOBAL}').combineWithMathJax({_: {`
+    ];
+    const packages = [];
+    PACKAGE = PACKAGE.sort(sortDir);
+    while (PACKAGE.length) {
+        const dir = path.dirname(PACKAGE[0]).split(path.sep)[0];
+        packages.push(processPackage(lines, '    ', dir));
+    }
+    lines.push('    ' + packages.join(',\n    '));
+    lines.push('}});');
+    fs.writeFileSync(path.join(LIB, COMPONENT + '.js'), lines.join('\n') + '\n');
+}
+
+/**
+ * Sort file paths alphabetically
+ */
+function sortDir(a, b) {
+    const A = a.replace(/\//g, '|');          // Replace directory separator by one that is after the alphnumerics
+    const B = b.replace(/\//g, '|');
+    return (A === B ? 0 : A < B ? -1 : 1);
+ }
+
+/**
+ * Recursively process packages, collecting them into groups by subdirectory, properly indented
```

Comment:
subdirectories

Reply:
The singular is correct, here.  For example, you collect marbles into groups by color, not by colors.